[ao] Implemented report generation for InputWeightEqualization Detector

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #80191

Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This includes the implementation and subsequent tests for the report
generation functionality of the detector. The full detector should now
be fleshed out and complete with this addition. This included
modifications to the ModelReportObserver class as well to capture min
and max per channel values. In addition, instead of passing in the
observer class to instantiate, the detectors now pass the ModelReport
instance the observer instance that they themselves instantiate.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Implemented report generation for InputWeightEqualization Detector

Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This includes the implementation and subsequent tests for the report
generation functionality of the detector. The full detector should now
be fleshed out and complete with this addition.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Implemented report generation for InputWeightEqualization Detector"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This includes the implementation and subsequent tests for the report
generation functionality of the detector. The full detector should now
be fleshed out and complete with this addition. This included
modifications to the ModelReportObserver class as well to capture min
and max per channel values. In addition, instead of passing in the
observer class to instantiate, the detectors now pass the ModelReport
instance the observer instance that they themselves instantiate.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -577,17 +600,214 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
                     "observer_args": targeted_node.args,
                 }
 
+                post_obs_fqn = fqn + "." + self.DEFAULT_POST_OBSERVER_NAME
```
don't we only need 1 observer? I'm unsure what the post observer is doing

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -577,17 +600,214 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+
+        # if global range is 0, throw error
+        if global_range == 0:
+            raise ValueError("The range of the info dict is 0")
```
i'd not reference the info dict in a user facing message, i.e. user has an input of all zeros, runs model report generator, we should probably give a better error message. Could just return None and handle it at upper level if that makes more sense.

HDCharles commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -577,17 +600,214 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
+
+            # raise error if not in weight info
+            if module_fqn not in weight_info:
+                raise KeyError("Both input_info and weight_info should have same keys")
```
again, could probably use better message to the user, i.e. could not find weight for module X rather than a technical error message.

vspenubarthi committed:
Update on "[ao] Implemented report generation for InputWeightEqualization Detector"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This includes the implementation and subsequent tests for the report
generation functionality of the detector. The full detector should now
be fleshed out and complete with this addition. This included
modifications to the ModelReportObserver class as well to capture min
and max per channel values. In addition, instead of passing in the
observer class to instantiate, the detectors now pass the ModelReport
instance the observer instance that they themselves instantiate.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

HDCharles commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1093,55 +1093,63 @@ def test_generate_report(self):
-                self.linear = torch.nn.Linear(3, 3)
-                self.relu = torch.nn.ReLU()
-                self.conv2d = torch.nn.Conv2d(3, 3, 1)
+    class LinearConv(torch.nn.Module):
```
potential nit: if you use similar things for many test cases, you can just have those defined outside the test class

HDCharles reviewed (changes requested):
needs more test cases for corner cases

what happens in no linear modules?

need to test fusion...etc

HDCharles reviewed (approved):
i'll accept subject to more test cases

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -588,6 +615,201 @@ def get_detector_name(self) -> str:
+            )
+
+            # if module is supported and it has a pre-observer
+            if is_supported_type and hasattr(module, self.DEFAULT_PRE_OBSERVER_NAME):
```
I'm seeing this structure repeated in a lot of places. I wonder if all detector classes should just define a `self.get_supported_modules(model)` method to minimize duplication

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -588,6 +615,201 @@ def get_detector_name(self) -> str:
 
+    def _extract_input_info(self, model: GraphModule) -> Dict[str, Dict]:
+        r"""
+        Takes in a callibrated GraphModule and then finds the relavent observers.
```
typo: relevant

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -607,7 +829,50 @@ def generate_detector_report(self, model: GraphModule) -> Tuple[str, Dict[str, A
+        no_use_str = "to not use"
+        input_weight_benefit_str = "we expect significant reduction in quantization error."
+        input_weight_non_benefit_reasoning = "the scales of the input vs. weight with regards to their ranges."
+        input_weight_non_benefit_str = "we don't expect much improvement from input-weight equalization based on {}"
```
Do we have any tests for these strings?

vspenubarthi replied:
There isn't a direct test for the strings themselves. However, all of them are formatted using values from the input_weight_equalization_info which is tested. The strings themselves are essentially just a boilerplate template that the dict populates, so the dict is what is primarily tested.

andrewor14 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -607,7 +829,50 @@ def generate_detector_report(self, model: GraphModule) -> Tuple[str, Dict[str, A
                 their s_c metric compared to the threshold
```
I can't comment up there but typo in L827: interest

andrewor14 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1160,3 +1168,59 @@ def forward(self, x):
+
+                    # make sure the global min and max were correctly recorded and presented
+                    self.assertEqual(input_range_info[InputWeightEqualizationDetector.GLOBAL_MIN_KEY], min(dimension_min))
+                    self.assertEqual(input_range_info[InputWeightEqualizationDetector.GLOBAL_MAX_KEY], max(dimension_max))
```
I feel there are a lot more things we can check here. E.g. Can we add a test for the comparison stat? We can assert whether it is in fact equal to the sqrt ratio given the input and weight ranges

andrewor14 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -30,9 +32,33 @@ def __init__(self):
+        x_copy = self._calculate_range_stats(x_copy)
+
+        # calculate the min, max stats
+        x_copy = self._calculate_min_max_stats(x_copy)
```
nit: These comments aren't helpful since they just repeat the method names. Best to remove them :)

andrewor14 commented on torch/ao/quantization/fx/_model_report/model_report_observer.py:
```diff
@@ -52,8 +78,41 @@ def forward(self, x):
+
+    def _calculate_min_max_stats(self, x_copy):
+        r"""
+        Args
```
nit: This docstring doesn't explain what this method does

andrewor14 reviewed (approved):
Looks great!

vspenubarthi committed:
Update on "[ao] Implemented report generation for InputWeightEqualization Detector"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This includes the implementation and subsequent tests for the report
generation functionality of the detector. The full detector should now
be fleshed out and complete with this addition. This included
modifications to the ModelReportObserver class as well to capture min
and max per channel values. In addition, instead of passing in the
observer class to instantiate, the detectors now pass the ModelReport
instance the observer instance that they themselves instantiate.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on test/quantization/fx/test_model_report_fx.py:
```diff
@@ -1179,3 +1200,118 @@ def test_input_weight_equalization_determine_points(self):
+                if "block1.linear" in module_fqn:
+                    block_1_lin_recs = input_weight_dict[module_fqn]
+                    # get input range info and the channel axis
+                    input_range_info = block_1_lin_recs[InputWeightEqualizationDetector.INPUT_INFO_KEY]
```
will users just be looking at the report or will they look at these informations as well? I think we should focus on testing what is used by user in the end

vspenubarthi replied:
This information is also passed as a dict object to the user, so they can perform their own analysis afterwards. Because the information is user facing, I thought it would be worth it to test.

jerryzh168 commented on torch/ao/quantization/fx/_model_report/detector.py:
```diff
@@ -306,12 +306,8 @@ def determine_observer_insert_points(self, prepared_fx_model: GraphModule) -> Di
-
-            if is_supported_type:
+            # make sure module is supported
+            if self._get_is_supported(module, insert=True):
```
nit: `_get_is_supported` --> `_is_supported`

vspenubarthi committed:
Update on "[ao] Implemented report generation for InputWeightEqualization Detector"


Summary: This adds the implementation for the InputWeightEqualization
detector. This includes both the implementation and the relavent test
cases. This detector is meant to be added to initialize a ModelReport
instance and it will keep track of the necessary statistics to decide if
for certain layers of interest (linear and conv for now), it makes sense
to use input weight equalization and gives the suggestion to the user.

This includes the implementation and subsequent tests for the report
generation functionality of the detector. The full detector should now
be fleshed out and complete with this addition. This included
modifications to the ModelReportObserver class as well to capture min
and max per channel values. In addition, instead of passing in the
observer class to instantiate, the detectors now pass the ModelReport
instance the observer instance that they themselves instantiate.

Test Plan: python test/test_quantization.py TestFxDetectInputWeightEqualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
